### PR TITLE
Update GetAllowedModifiers

### PR DESCRIPTION
**Pull request by chborl**
Fixes https://github.com/dotnet/roslyn/issues/32423

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 1]
```diff
@@ -1523,17 +1523,68 @@ public override SyntaxNode WithAccessibility(SyntaxNode declaration, Accessibili
...
+            DeclarationModifiers.ReadOnly;
+        private static readonly DeclarationModifiers s_interfaceModifiers =
+            DeclarationModifiers.New |
+            DeclarationModifiers.Partial;
+        private static readonly DeclarationModifiers s_accessorModifiers =
+            DeclarationModifiers.New;
+        private static readonly DeclarationModifiers s_localDeclarationModifiers =
+            DeclarationModifiers.Ref;
+        private static readonly DeclarationModifiers s_parameterModifiers =
+            DeclarationModifiers.Ref;
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Looking at [DeclarationModifiers](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Editing/DeclarationModifiers.cs,63) and its properties, it seems it is designed only for top level declarations - i.e., namespaces, types, methods, fields, properties, and events, but not for other symbols such as locals or parameters. Its not good that there are absolutely no comments or even doc comments on the type, especially given its a public API.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 2]
```diff
@@ -1523,17 +1523,68 @@ public override SyntaxNode WithAccessibility(SyntaxNode declaration, Accessibili
...
+            DeclarationModifiers.New |
+            DeclarationModifiers.ReadOnly |
+            DeclarationModifiers.Static;
+        private static readonly DeclarationModifiers s_methodModifiers =
+            DeclarationModifiers.Abstract |
+            DeclarationModifiers.Async |
+            DeclarationModifiers.New |
+            DeclarationModifiers.Override |
+            DeclarationModifiers.Partial |
+            DeclarationModifiers.ReadOnly |
```
Comment (mavasani) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Methods cannot be readonly

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 3]
```diff
@@ -1523,17 +1523,68 @@ public override SyntaxNode WithAccessibility(SyntaxNode declaration, Accessibili
...
+            DeclarationModifiers.Sealed |
+            DeclarationModifiers.Static |
+            DeclarationModifiers.Virtual;
+        private static readonly DeclarationModifiers s_constructorModifiers =
+            DeclarationModifiers.Static;
+        private static readonly DeclarationModifiers s_propertyModifiers =
+            DeclarationModifiers.Abstract |
+            DeclarationModifiers.New |
+            DeclarationModifiers.Override |
+            DeclarationModifiers.ReadOnly |
```
Comment (mavasani) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Properties cannot be readonly

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 4]
```diff
@@ -1523,17 +1523,68 @@ public override SyntaxNode WithAccessibility(SyntaxNode declaration, Accessibili
...
-        private static readonly DeclarationModifiers s_eventFieldModifiers = DeclarationModifiers.New | DeclarationModifiers.Static;
-        private static readonly DeclarationModifiers s_indexerModifiers = DeclarationModifiers.Abstract | DeclarationModifiers.New | DeclarationModifiers.Override | DeclarationModifiers.ReadOnly | DeclarationModifiers.Sealed | DeclarationModifiers.Static | DeclarationModifiers.Virtual;
-        private static readonly DeclarationModifiers s_classModifiers = DeclarationModifiers.Abstract | DeclarationModifiers.New | DeclarationModifiers.Partial | DeclarationModifiers.Sealed | DeclarationModifiers.Static;
-        private static readonly DeclarationModifiers s_structModifiers = DeclarationModifiers.New | DeclarationModifiers.Partial;
-        private static readonly DeclarationModifiers s_interfaceModifiers = DeclarationModifiers.New | DeclarationModifiers.Partial;
-        private static readonly DeclarationModifiers s_accessorModifiers = DeclarationModifiers.Abstract | DeclarationModifiers.New | DeclarationModifiers.Override | DeclarationModifiers.Virtual;
+        private static readonly DeclarationModifiers s_fieldModifiers =
```
Comment (mavasani) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Is it possible to separate out the formatting change into a separate commit? That way it is easier to see what changed here. Thanks!

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 5]
```diff
@@ -1583,9 +1634,13 @@ private static DeclarationModifiers GetAllowedModifiers(SyntaxKind kind)
                 case SyntaxKind.RemoveAccessorDeclaration:
                     return s_accessorModifiers;
 
-                case SyntaxKind.EnumMemberDeclaration:
-                case SyntaxKind.Parameter:
                 case SyntaxKind.LocalDeclarationStatement:
+                    return s_localDeclarationModifiers;
```
Comment (mavasani) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: I believe this change should be reverted - in fact I wouldn't even list parameter, local declaration in this switch and just let it fall through to default case.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: If we are planning to extend `DeclarationModifiers` to now include modifiers for non-top level declarations, then we need to extend it to include modifiers such as `out`, `in`, `volatilte`, etc. - we need a more exhaustive list. Additionally, this needs to go through a design meeting as we are extending the scope of an existing public API (although completely undocumented).

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: properties can be readonly in VB :)

Comment (mavasani) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Ah, VB :-) Thanks!

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: > If we are planning to extend DeclarationModifiers to now include modifiers for non-top level declarations

We shoudl ask @mattwar .  But my remembrance is that this API was intended for use for all the use-cases of SyntaxGenerator.  In that regard, it would be expected to be meaningful on all these other constructs as well.  It's just the case that we often have shipped stuff that hit the 95% case, and allowed the rest to come in through bug fixes as features have needed additional functionality.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: i personally disagree.  SyntaxGenerator attempts to be a 'smoothing' out API, where one can operate on any nodes where a *sensible* interpretation can be provided.  In this case, Locals can def have modifiers, so it seems reasonable to support this.

I'd def like to hear @mattwar 's thoughts though as maybe i'm entirely offbase and he strongly felt these shoudl be out of scope of SyntaxGenerator.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Seems reasonable to extend - I'd still prefer it go through a design meeting and also we add the exhaustive list at least for new symbol kinds we are adding (parameters, locals). @chborl once finalize, can you please add doc comments to all the public APIs in this type.

**Comment by CyrusNajmabadi** [comment]
To expand more on: 
>  SyntaxGenerator attempts to be a 'smoothing' out API

We def treat raw 'Syntax' as a layer where we make everything explicit and force users to have to understand and work with each node type separately.  We're strict about everything, and we punish if you violate invariants (which i think is a good thing!).

SyntaxGenerator is the contrasting API to that.  It is far more relaxed, and takes an approach of trying to do "the right-ish enough" thing as often as it can.  It both works over VB/C#, allowing you to ignore the differences for parts of the language that are 'close enough', and it wants to let you do things that "make sense-ish" across many different nodes in the same language.  For example, unlike 'Syntax', 'SyntaxGnerator' has a nice and simple way to say "give me all the attributes on this node".  This is a sensible operation across VB/C#, and across many nodes in C# and VB.  If new pieces of syntax were added that supported attributes, we'd def expand SyntaxGenerator to support those.

Another example of 'SyntaxGEnerator' doing 'rightish' enough is how it handles moving members around.  If, for example, you move members into an interface, then it will automatically remove hte body as well as the modifiers.  This is because it can sensibly judge that this is most likely what you wanted, while allowing you to make the syntactic change explicitly if you don't.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: > I'd still prefer it go through a design meeting

Totally agree.  Public API.  So it def should be discussed.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 3 above)
Comment (Neme12) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: But this is inside CSharpSyntaxGenerator. C# properties cannot be readonly.

**Comment by sharwell** [comment]
Design review conclusion: we should update syntax generator to correctly handle all supported modifiers for symbols that can appear in a signature. This involves two specific items:

1. Ensuring the supported modifiers are correct for symbol types that are already supported (e.g. `extern` should be supported for methods)
2. Adding support for parameters and type parameters, which are not top-level symbols but are visible in the signatures of top-level symbols

Since this API does not currently support generating method bodies in a language-agnostic way, we do not need to support modifiers for items that can only appear in these locations (e.g. locals and local functions).

**Comment by CyrusNajmabadi** [comment]
> Since this API does not currently support generating method bodies in a language-agnostic way, we do not need to support modifiers for items that can only appear in these locations (e.g. locals and local functions).

Could you clarify this?  The API *does* at least support generating statements, and things like local declarations.  So it seems sensible that modifiers would work there.  Thoughts?
